atlp-rwanda/stackup2-barefoot-backend
# #171380755 Email and in-app notifications

#### What does this PR do?

- It provides both in-app and Email notifications upon user create trip request

#### Description of Task to be completed?

- When a user creates a trip request they get an email notification about their trip request.

- A user can also disable email notifications so that they don't receive email notifications anymore.

- A manager is able to get the in-app notifications.

#### How should this be manually tested?

- `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend.git`
- `cd stackup2-barefoot-backend`
- `git fetch origin ft-email-inapp-notifications-171380755`
-`git checkout ft-email-inapp-notifications-171380755`
- `npm install`
- Use `.env.example` as reference for environment variables
- In case you don't have a functional database for this project, run `npm run create-db`
- `npm run migrations`
- `npm run dev`
- At this point, the server must be running and the application should be connected to the database and all the tables in the database are created.
- Signup using this URL `localhost:3000/api/auth/signup` use this sample object 
```
{
	"firstName": "YOUR_FIRST_NAME",
	"lastName":"YOUR_LAST_NAME",
	"address":"YOUR_ADDRESS",
	"username":"YOUR_USERNAME",
	"password":"YOUR_PASSWORD",
	"email":"YOUR_EMAIL",
	"gender":"YOUR_GENDER"
}
```
- Login in order to verify your account using this URL `localhost:3000/api/auth/login` use this sample object

```
"email":"YOUR_EMAIL",
"password":"YOUR_PASSWORD"

```
- create a trip using this URL `localhost:3000/api/trips` use this object sample 
```
{
    "travelFrom": "kigali",
    "travelTo": "Butare",
    "travelDate": "2020-05-14",
    "travelReason": "business meeting",
    "travelType": "one-way",
    "accommodation": true
}

```

- Use this URL to disable email notifications, PATCH: `http://localhost:3000/api/auth/disable?mode=email`

- Use this URL to disable in-appnotifications, PATCH: `http://localhost:3000/api/auth/disableOrEnable?mode=inapp`

-  Use this URL to get in-app notifications `localhost:3000/api/trips/inapp/notification/`
- Use this URL to get unread in-app notifications `localhost:3000/api/trips/inapp/notification?status=unread`

#### Any background context you want to provide?

- When you signup please provide a real email address that you can access because notifications will be sent that email address that you provided during signup.

- In order to get in-app notifications you have to be a manager.

#### What are the relevant pivotal tracker stories?

- [171380755](https://www.pivotaltracker.com/story/show/171380755)

#### Screenshots (if appropriate)
![inapp3](https://user-images.githubusercontent.com/37122177/81540666-4451fc80-9372-11ea-8f03-30e61648c7a7.PNG)
![inapp2](https://user-images.githubusercontent.com/37122177/81540676-46b45680-9372-11ea-8eb6-7ddd55454e4e.PNG)
![inapp1](https://user-images.githubusercontent.com/37122177/81540677-47e58380-9372-11ea-9651-008909963853.PNG)
![requesterEmail](https://user-images.githubusercontent.com/37122177/81540679-487e1a00-9372-11ea-98ca-680fd060d83c.PNG)
![managerEmail](https://user-images.githubusercontent.com/37122177/81540683-4916b080-9372-11ea-828a-b0efdb2cb372.PNG)
![nonManager](https://user-images.githubusercontent.com/37122177/81540687-49af4700-9372-11ea-86b6-873421f5109c.PNG)
![managerDontExist](https://user-images.githubusercontent.com/37122177/81540691-4ae07400-9372-11ea-9d5e-eacdcba9dd5a.PNG)
